rust/src/cli/persist_nic.rs

```diff
@@ -195,49 +116,146 @@ pub(crate) fn run_persist_immediately(
-        );
-        Ok(())
+    for (iface_name, file_path) in pinned_ifaces.drain() {
+        if !is_nmstate_generated_systemd_link_file(&file_path) {
+            log::info!(
+                "File {} is not generated by nmstate, ignoring",
+                file_path.display()
+            );
+            continue;
+        }
+        let systemd_iface_name =
+            match get_systemd_preferred_iface_name(&iface_name) {
+                Ok(i) => i,
+                Err(e) => {
+                    log::error!(
+                        "Failed to retrieve systemd preferred \
+                        iface name for {iface_name}: {e}"
+                    );
+                    continue;
+                }
+            };
+        if systemd_iface_name == iface_name {
+            if dry_run {
+                log::info!(
+                    "The generated {} file has no effect for \
+                    interface {iface_name}, will remove if not \
+                    in dry-run mode",
+                    file_path.display()
+                );
+            } else {
+                log::info!(
+                    "The generated {} file has no effect for \
+                    interface {iface_name}, removing",
+                    file_path.display()
+                );
+                std::fs::remove_file(file_path)?;
+            }
+        } else {
+            log::info!(
+                "Systemd generate interface name \
+                '{systemd_iface_name}' != pinned name '{iface_name}', \
+                will keep config file {}",
+                file_path.display()
+            );
+        }
+    }
+    Ok("".to_string())
+}
+
+// With `NamePolicy=keep kernel database onboard slot path` in systemd configure
+// in RHEL 8 and 9. Assuming `keep, kernel and database` all return NULL,
+// Systemd will use interface name in the order of:
+//  * `ID_NET_NAME_ONBOARD`
+//  * `ID_NET_NAME_SLOT`
+//  * `ID_NET_NAME_PATH`
+pub(crate) fn get_systemd_preferred_iface_name(
+    iface_name: &str,
+) -> Result<String, CliError> {
+    let output: String = String::from_utf8(
+        std::process::Command::new("udevadm")
+            .args(UDEVADM_CMD_OPT)
+            .arg(&format!("/sys/class/net/{iface_name}"))
+            .output()?
+            .stdout,
+    )
+    .map_err(|e| {
+        CliError::from(format!("Failed to parse udevadm reply to UTF-8: {e}"))
     })?;
 
-    Ok("".to_string())
+    for line in output.split('\n') {
+        if let Some(n) = line.strip_prefix(&format!("{}=", ID_NET_NAME_ONBOARD))
+        {
+            return Ok(n.to_string());
+        }
+    }
+
+    for line in output.split('\n') {
+        if let Some(n) = line.strip_prefix(&format!("{}=", ID_NET_NAME_SLOT)) {
+            return Ok(n.to_string());
+        }
+    }
+
+    for line in output.split('\n') {
+        if let Some(n) = line.strip_prefix(&format!("{}=", ID_NET_NAME_PATH)) {
+            return Ok(n.to_string());
+        }
+    }
```

Comment:
the `fold` function cause the iterator goes through the whole output, while my code return the value once ideal value found. So the performance difference need to debate. 

And the `fold` code is not rust-learner friendly. It looks more like perl to me.

Reply:
It's a bit subtle, because in practice we can only skip iterating the whole output if we find `ID_NET_NAME_ONBOARD`.  If we find `ID_NET_NAME_SLOT` we still need to finish in case `ID_NET_NAME_ONBOARD` will come after.

I don't have a strong opinion here; but the string-allocation in a loop seems unnecessary.
How about https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=7ba83c816fabe298655742c9766122b3 ?

Reply:
Adopted the new `split_once()` code.